subprojects/core/src/main/java/org/gradle/api/internal/tasks/userinput/AbstractUserInputHandler.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.userinput;
+
+import org.gradle.api.internal.provider.Providers;
+import org.gradle.api.provider.Provider;
+import org.gradle.internal.Try;
+
+import java.util.function.Function;
+
+public abstract class AbstractUserInputHandler implements UserInputHandler {
+    @Override
+    public <T> Provider<T> askUser(Function<? super UserQuestions, ? extends T> interaction) {
+        return Providers.changing(new Providers.SerializableCallable<T>() {
+            Try<T> result;
+
+            @Override
+            public T call() {
+                if (result == null) {
+                    try {
+                        try (CloseableUserQuestions userQuestions = AbstractUserInputHandler.this.newInteraction()) {
```

Comment:
🤔 If we need to wrap the `AutoClosable` in an outer `try`, its not doing much work.

I think this should either 1) get rid of the outer `try`, or 2) consider removing the `CloseableUsrQuestions` class entirely in favor of just using `UserQuestions` and manually calling the `close` logic in a new `finally` block here.

My argument for doing 2) is that: it removes an unused `Closeable` type, the `NonInteractive` implementation has nothing to close, and the `Default` implementation isn't doing anything that seems much like _closing_ - it's conditionally broadcasting an event.  Should this event even be broadcast if `result instanceof Failure` here?